'''
##############################################################################
# Mapping using minimap2
##############################################################################

This rule will map fastq file for a single sample and batch to the reference. 

Output will be stored in the "mapping" directory as a bam file including a 
bai index file.

The following steps are performed:
  - Alignment
  - Filter only unique mappings
  - Sort bam file
  - Create bam index
'''
rule alignment_chr11_chr17:
    input: rules.fast5_to_fastq.output.fq
    output: os.path.join(basedir,'mapping_chr11_chr17/{sample}/{batch}.sorted.filtered.bam')
    params:
        jobname='align_{sample}_{batch}',
        runtime='02:00',
        memusage='32000',
        slots='8',
        misc=''
    shell: '''
           {minimap2} -t {params.slots} -a -x map-ont {reference_chr11_chr17} {input} | {samtools} view -q 1 -b | {samtools} sort -@ {params.slots} -T {output}.tmp -o {output}
           {samtools} index {output}
           '''

rule all_alignment_chr11_chr17:
    input: expand(rules.alignment_chr11_chr17.output, zip_combinator, sample=sbf.sb_samples, batch=sbf.sb_batches)

